### ISSUE TITLE
[Feature Request] TimeSpan.Dehumanize()

### ISSUE THREAD

**Issue by quole** [Humanizr/Humanizer#691]
Yes, I know there's been some lengthy discussion about dehumanizing dates (#112). And the [readme](https://github.com/Humanizr/Humanizer/blob/dev/readme.md) even states: 

> No dehumanization for dates as `Humanize` is a lossy transformation and the human friendly date is not reversible"

There's also already a couple of libraries that deal with ambiguous expressions such as "yesterday", "3 weeks ago", and "7 hours before tomorrow at midnight". e.g. [nChronic](https://github.com/robertwilczynski/nChronic/) and Microsoft LUIS.

However, I can't see any discussion about dehumanizing `TimeSpan`s, which are much more straight forward. Even if the main focus here is on humanizing rather than the reverse, this kind of thing would be well suited to this library

## TimeSpan

`TimeSpan`s are much less ambiguous than dates. There also don't seem to be many other libraries that deal with them well. Here's some basic examples / semi-pseudocode tests for a TimeSpan dehumanizer: 

```
var timeTest = new TimeSpan(3, 18, 0)
Assert.Equal(TimeSpan.Dehumanize("three hours and eighteen minutes"), timeTest);
Assert.Equal(TimeSpan.Dehumanize("3 hours, 18 minutes"), timeTest);
Assert.Equal(TimeSpan.Dehumanize("3:18 hours"), timeTest);
Assert.Equal(TimeSpan.Dehumanize("3h18m"), timeTest);
Assert.Equal(TimeSpan.Dehumanize("3h 18m"), timeTest);
Assert.Equal(TimeSpan.Dehumanize("3.3hrs"), timeTest);
Assert.Equal(TimeSpan.Dehumanize("3:18:00"), timeTest);
Assert.Equal(TimeSpan.Dehumanize("3:18"), timeTest);
Assert.Equal(TimeSpan.Dehumanize("18:01", expectedUnits:minutes), new TimeSpan(0, 18, 01);
```

I think this kind of thing would be fairly straight forward and without ambiguity.

The other libraries I've found that deal with dehumanizing TimeSpans are: 
* Microsoft LUIS, e.g. [`builtin.datetimeV2.duration`](https://docs.microsoft.com/en-us/azure/cognitive-services/luis/luis-reference-prebuilt-entities#builtindatetimev2) but it seems like it's overly complex and I'm guessing it's not open source. Also seems more focused on speech than text.
* [Exceptionless.DateTimeExtensions](https://github.com/exceptionless/Exceptionless.DateTimeExtensions#timeunit) (Apache 2.0 licensed) has a parser that works as following: `TimeSpan oneHour = TimeUnit.Parse("3h");` But would fail on all of the example strings I've given above as they are "complex".

TL;DR: Humanizer would be a good library to add a TimeSpan Dehumanizer to. I could be wrong, but I can't see anything overly complicated about it (unlike dehumanizing TimeDate text), as long as you stay away from accepting highly ambiguous inputs like "2 months" and stick to making it work as would be expected. I don't know much about what sort of things might trip up localization efforts though, but making it part of Humanizer would give a good chance for localization to occur.

I haven't (yet) poked around with the inner workers of Humanizer's parsers to see how easy it might be to add something like this, but thought I'd put it out there to see what interest was like.

**Comment by onovotny** [comment]
I'm curious about what context you would see this being useful?

**Comment by quole** [comment]
@onovotny My use case is a Discord (chat) bot which can act as a timer for "writing sprints" (a glorified 5 to 45 minute timer). So it accepts commands from members of a chat such as:

`!sprint 15` (to run a timer for 15 minutes)

example:

![image](https://user-images.githubusercontent.com/800133/36929167-59d1fb1c-1ee1-11e8-808e-ca1b873a85c2.png)

`!sprint 1000s` works thanks to DateTimeExtensions mentioned in the OP.

But ideally it would accept more arbitrary input like:

`!sprint 12m 30s`
`!sprint 0.25 hrs` (this one doesn't for multiple reasons: floating point, space, odd spelling of hours)
`!sprint 15:10` (treated as 15h10s by most parsers, aka `15:10:00`. However a parameter to the parser should get this parsed as `00:15:10`)
etc

It's not 100% required feature, but I was surprised that there wasn't a library that enabled it easily already (that I can find)

So basically, the use case is for entering a length of time for a timer where only text input is available, and the users are generally not be technical and have a low tolerance for failure (every failed attempt is seen by all members of the chat).

More generally, I'd imagine this sort of function would also be useful anywhere users need to enter a time period as text, including web forms and command line arguments.

**Comment by ericmck2000** [comment]
I will +1 this feature, for nearly the exact same use-case (Discord Bot)

**Comment by ericmck2000** [comment]
However, I will also note, The discord.net library already has built in support for this, using the TimeSpan type converter.
https://github.com/RogueException/Discord.Net

https://discord.foxbot.me/docs/guides/commands/commands.html#type-readers

**Comment by quole** [comment]
I had a go at creating this a while ago but never got around to polishing it up to share. I'll upload it as a separate github project when I get a chance. It allows inputs like "3m20s" or "3 min 20.5secs" or "3:00:10" etc. 

I didn't know about Discord.Net's parser. I'll put in some tests to check mine is compatible with everything it accepts too.

**Comment by ericmck2000** [comment]
I created a pull request for discord.net's version to fall back and use the
built in parse method which helps alot.

It's not perfect, but, it understands...
1:10:00,
6d,
10m,
3s, etc....

**Comment by ericmck2000** [comment]
https://github.com/RogueException/Discord.Net/pull/1131

So, to summarize, the discord.net version, will soon just be mostly utilizing the builtin timespan.parse methods.

It would be really great for humanizer to support this feature!

**Comment by quole** [comment]
Okay, here's my TimeSpanParser:

https://github.com/quole/TimeSpanParser

It doesn't parse "three hours and eighteen minutes" but it will do the rest of the examples.

**Comment by icnocop** [comment]
I'm using https://github.com/Pathoschild/NaturalTimeParser.

Even though it hasn't been updated for some time, it works well for my needs.

**Comment by bernatgy** [comment]
I'd like to see something like this as well... My use case is the following;

I'm trying to achieve user friendly configuration file for one of my app's timer based features. I was already using Humanizer to save a default in the format "x minutes, y seconds", by the time I noticed there's no Dehumanize for those values. One would expect something at least for this default format. 😢

**Comment by prodigy1** [comment]
I support this as well, my use case is configuration of an application. As a TimeSpan cannot properly and reliable contain a span of months, but a string like "3 months" can, I would like to request this as a feature - together with a starting point from a given date. In that case you can correctly "dehumanize" the string "3 months" to a specific amount of days.

**Comment by fdcastel** [comment]
I have a similar need for configuration files. 

Instead of 

timeout: 00:00:00.050

I could use

timeout: 50ms

or 

```
timeout: 50 milliseconds

**Comment by pengowray** [comment]
@fdcastel [TimeSpanParser](https://github.com/pengowray/TimeSpanParser) will parse all those values to a TimeSpan. e.g. `TimeSpanParser.Parse("50 milliseconds")`

Note: TimeSpanParser only does English, and it will **not** do the relative dates mentioned by others such as "3 months from today" (though it will handle the unambiguous "0 months"). I've considered adding relative dates but I don't personally have a need for them.

I'd still love if Humanizer added this functionality. I have a lot of unit tests you guys could use. They were more comprehensive than Microsoft's and I even found a bug in Microsoft's `TimeSpan.Parse` code that I got patched.

**Comment by caunt** [comment]
Sad finding this not implemented yet

**Comment by wyf027** [comment]
Opened **#1800** with an MVP parser for compact duration strings (3h18m, 12m 30s, colon forms). Word-based and localized parsing left for follow-up.

**Comment by wyf027** [comment]
Follow-up **#1801** adds word-phrase parsing (comma/and separators) and round-trip tests against `TimeSpan.Humanize()`. Depends on **#1800**. Both CI green + approved.

**Comment by wyf027** [comment]
**Status (2026-05-27):** MVP in **#1800** (compact + colon, CI green, APPROVED). Word-phrase + round-trip in **#1801** (stacked, APPROVED). Localized phrases / `toWords` / months remain follow-ups after merge.

**Comment by clairernovotny** [comment]
Implemented and merged in #1854. The first supported contract is deliberately invariant and deterministic:

- standard `TimeSpan` forms use the BCL's invariant parser and colon semantics;
- compact tokens support exact `ms`, `s`, `m`, `h`, `d`, and `w` values, with one week equal to seven days;
- both throwing and `Try` APIs are available.

Localized phrases, number words, alternate colon interpretations, months, and years remain outside this first slice. The APIs will be available in the next Humanizer release containing #1854.

Thanks again to @pengowray for the original report and examples, and to @wyf027 for the earlier exploration in #1800 and #1801 that helped shape the compatible API surface and edge-case coverage.
